File: examples/hello.kt
import java.util.Scanner

internal abstract class Hello<T>(count: T) {
    fun sum(): T {
        return genericSum(zero, ::plus)
    }

    fun product(): T {
        return genericSum(one, ::multiplies)
    }

    protected abstract val zero: T
    protected abstract val one: T
    protected abstract fun getSequence(count: T): Iterable<T>
    protected abstract fun plus(a: T, b: T): T
    protected abstract fun multiplies(a: T, b: T): T

    private inline fun genericSum(identity: T, operation: (T, T) -> T): T {
        return sequence.fold(identity, operation)
    }

    private val sequence: Iterable<T> = this.getSequence(count)
}

fun main() {
    val n = Scanner(System.`in`).nextLong()
    val hello = object : Hello<Long>(n) {
        override val zero = 0L
        override val one = 1L
        override fun getSequence(count: Long): Iterable<Long> {
            return (1..n).asIterable()
        }
        override fun plus(a: Long, b: Long): Long {
            return a + b
        }
        override fun multiplies(a: Long, b: Long): Long {
            return a * b
        }
    }
    println(hello.sum())
    println(hello.product())
}

File: examples/hello.dart
import 'dart:io';

abstract class Hello<T> {
  Hello(T count) {
    _sequence = getSequence(count);
  }

  T sum() => _genericSum(zero, plus);
  T product() => _genericSum(one, multiplies);

  T get zero;
  T get one;
  Iterable<T> getSequence(T count);
  T plus(T a, T b);
  T multiplies(T a, T b);

  late final Iterable<T> _sequence;
  T _genericSum(T identity, T Function(T, T) operation) =>
      _sequence.fold(identity, operation);
}

class _HelloBigInt extends Hello<BigInt> {
  _HelloBigInt(super.count);

  @override
  BigInt get zero => BigInt.zero;

  @override
  BigInt get one => BigInt.one;

  @override
  Iterable<BigInt> getSequence(BigInt count) sync* {
    var i = BigInt.one;
    while (i <= count) {
      yield i;
      i += BigInt.one;
    }
  }

  @override
  BigInt multiplies(BigInt a, BigInt b) => a * b;

  @override
  BigInt plus(BigInt a, BigInt b) => a + b;
}

void main() {
  final n = BigInt.parse(stdin.readLineSync()!);
  final hello = _HelloBigInt(n);
  print(hello.sum());
  print(hello.product());
}